SoD proposal/doc update.

hjoliver opened the pull request:
Two new commits.
- The first commit documents exactly how suicide triggers work under SoD (after an email from @dpmatthews)
- The second commit documents the new handling of absolute triggers (after SoD review comments from @dwsutherland)

Both add minor items to the TODO list, to be converted to issues after the SoD PR is merged.

wxtim commented on docs/proposal-spawn-on-d.md:
```diff
@@ -205,22 +205,51 @@ trigger events too, but for now a waiting task with an xtrigger is fine).
+feasible.
+
 (Note without the first line above `start` is not defined on any sequence, so
 with SoS the scheduler will stall with `foo` unsatisfied, but with SoD it will
 shut down immediately with nothing to do because and `foo` never gets spawned.)
```
```suggestion
with SoS the scheduler will stall with `bar` unsatisfied, but with SoD it will
shut down immediately with nothing to do because and `bar` never gets spawned.)
```

hjoliver replied:
Good spotting @wxtim . I had changed `foo` to `bar` in the example, to be more like the new example added below it :grimacing:

dpmatthews commented on docs/proposal-spawn-on-d.md:
```diff
@@ -205,22 +205,51 @@ trigger events too, but for now a waiting task with an xtrigger is fine).
+
 (Note without the first line above `start` is not defined on any sequence, so
 with SoS the scheduler will stall with `foo` unsatisfied, but with SoD it will
 shut down immediately with nothing to do because and `foo` never gets spawned.)
```
"because and" => "because"
`foo` should be `bar` (and 2 lines above)

dpmatthews replied:
Updated version still says "bacause and" ?

hjoliver replied:
Fixed.

oliver-sanders commented:
Seems legit, absolute dependencies should be fairly few (usually just one) so the overheads of this caching are pretty minimal.

dpmatthews commented on docs/proposal-spawn-on-d.md:
```diff
@@ -205,22 +205,51 @@ trigger events too, but for now a waiting task with an xtrigger is fine).
+pool spawn its next instance (strictly speaking this is only necessary for
+tasks with no non-absolute triggers to spawn them on demand).
+
+#### Retriggering
```
"retrigger the absolute parent without reflow, then retrigger with reflow the first child that you want to continue from"
That will only run that single child won't it?

hjoliver replied:
- "retrigger the absolute parent without reflow" - this will only trigger the absolute parent task
- "retrigger with reflow the first child that you want to continue from" - this will trigger "the first child that you want to continue from" and that will (on release from the runahead pool) spawn the next child, and so on - because that's how we spawn tasks with absolute parents. (Note I've said "spawn" rather than "run" for subsequent child tasks).

hjoliver replied:
I'll make the text clearer...

hjoliver commented:
Comments addressed.

hjoliver commented:
> Seems legit, absolute dependencies should be fairly few (usually just one) so the overheads of this caching are pretty minimal.

Caching the trigger results is the SoD version of keeping absolute parents in the task pool forever in SoS.

hjoliver commented:
Having addressed the comments and gotten one approval I'm going to merge this so that the document is up to date with the just-merged implementation PR. (We can always tweak the docs again later).
